e[quantization] TorchScript Serialization for dynamic LSTM module

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25877 e[quantization] TorchScript Serialization for dynamic LSTM module**
* #25870 [quantzation] Test scripting and tracing for dynamic linear modules
* #25869 Fix c10 tracing

Differential Revision: [D17275746](https://our.internmc.facebook.com/intern/diff/D17275746)

Commit:
[quantization] Test tracing and scripting for dynamic LSTM module

Commit:
Update on "[quantization] Test tracing and scripting for dynamic LSTM module"

Commit:
Update on "[quantization] Test tracing and scripting for dynamic LSTM module"

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -55,9 +50,7 @@ def __init__(self, mode, input_size, hidden_size,
             raise ValueError("Unrecognized RNN mode: " + mode)
 
         self._all_weights = []
```
Maybe we can change `_all_weights` to `_all_weights_names`? This is more consistent with `_all_weight_values`.

jianyuh commented on test/test_quantization.py:
```diff
@@ -519,8 +519,32 @@ def test_quantized_rnn(self):
+            torch.testing.assert_allclose(out_val, ref_val)
+
+        # Test save/load
+        b = io.BytesIO()
```
nit: should we use the same tmpFile way (`with tempfile.TemporaryFile() as f:`) for testing?
https://github.com/pytorch/pytorch/blob/master/test/test_quantized_nn_mods.py#L133

I am OK if you think `io.BytesIO` is more convenient. Just want to make it consistent.

jamesr66a replied:
No, I think we should move more of these toward using BytesIO. Using files in tests is generally considered bad practice because it introduces syscalls and dependencies on the execution environment, and thus can cause test flakiness/instability

jianyuh replied:
Make sense. I guess it would be great to change all `tempfile` to `io.BytesIO` then.

jianyuh replied:
Following this issue, creating a PR to fix the quantization related test: https://github.com/pytorch/pytorch/pull/25976

Commit:
Update on "[quantization] TorchScript Serialization for dynamic LSTM module"


Differential Revision: [D17275746](https://our.internmc.facebook.com/intern/diff/D17275746)

jianyuh reviewed (approved):
LGTM!

jamesr66a commented:
@jianyuh We actually found a bug here that wasn't caught by the tests: __setstate__ doesn't pack bias together with the weights like it's supposed to. I'm gonna try to repro and fix it in this PR

Commit:
Update on "e[quantization] TorchScript Serialization for dynamic LSTM module"


Differential Revision: [D17275746](https://our.internmc.facebook.com/intern/diff/D17275746)

jamesr66a commented:
@jianyuh I updated this diff, PTAL

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -278,10 +280,14 @@ static std::vector<QuantizedCellParamsDynamic> gather_quantized_params_dynamic(
-  for (size_t i = 0; i < params.size(); i += 4) {
-    result.emplace_back(params[i], params[i + 1], params[i + 2], params[i + 3]);
+  for (size_t i = 0; i < params.size(); i += 2) {
+    auto& packed_struct_ih = cpp_custom_type_hack::cast<PackedLinearWeight>(params[i]);
```
Check the OSS test failure. Need to add `#ifdef USE_FBGEMM`.

jianyuh replied:
I think so. The alternative is to provide a general C/C++ reference slow implementation if USE_FBGEMM is undefined. But as you mentioned before, it would be better to force the error here and tell the users this is not correct way. The similar handler is here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear.cpp#L224-L236

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -54,31 +49,24 @@ def __init__(self, mode, input_size, hidden_size,
                     # for each layer, for each direction we need to quantize and pack
                     # weights and pack parameters in this order:
                     #
                     #   w_ih, w_hh, b_ih, b_hh
```
Nit: Remove b_ih, b_hh.

jianyuh commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -224,28 +228,20 @@ def process_weights(ihhh, layer, suffix):
-                    params = [packed_weight, bias]
-                    pos_names = ['w', 'b']
+                    params = [packed_weight]
+                    pos_names = ['w']
```
nit: in the comment of Line 223, remove `b_ih, b_hh`.

jianyuh replied:
Also remove Line 212 and 215.

jianyuh reviewed (commented):
> @jianyuh We actually found a bug here that wasn't caught by the tests: **setstate** doesn't pack bias together with the weights like it's supposed to. I'm gonna try to repro and fix it in this PR

Sorry for not catching this. I think this is after the PR: https://github.com/pytorch/pytorch/pull/25428, as the bias is temporarily stored in `PackedLinearWeight` for solving the dynamic batching issues in the eager mode.

Commit:
Update on "e[quantization] TorchScript Serialization for dynamic LSTM module"


Differential Revision: [D17275746](https://our.internmc.facebook.com/intern/diff/D17275746)

Commit:
Update on "e[quantization] TorchScript Serialization for dynamic LSTM module"


Differential Revision: [D17275746](https://our.internmc.facebook.com/intern/diff/D17275746)

jianyuh commented on test/test_quantization.py:
```diff
@@ -519,8 +519,34 @@ def test_quantized_rnn(self):
+        for out_val, ref_val in zip(final_hiddens_int8, ref_hid):
+            torch.testing.assert_allclose(out_val, ref_val)
+
+        class ScriptWrapper(torch.nn.Module):
```
Can we reuse the `model` above as `model = LSTMDynamicModel().eval()`, instead of creating a new PyTorch module wrapper? It is defined in https://github.com/pytorch/pytorch/blob/master/test/common_quantization.py#L182-L190. I am OK with the current code if you think this is more consistent in the unit test.

jamesr66a replied:
No, this is to ensure that the serialization and scripting functionality works even in a nested context. This is a case that has often been overlooked in the past
